proposals/NNNN-improve-starting-streaming-services.md

```diff
@@ -0,0 +1,54 @@
+
+## Potential downsides
+
+In certain situations where the retry mechanism occurs, the proposed solution can increase the time until the app has started the video and audio stream for up to two seconds. However, the libraries gain a lot of stability for error cases which is much more valuable.
+
+It can be possible that this retry mechanism puts some stress to existing infotainment systems as they were never tested against this method. However, the current behavior of the libraries to stop after one failure would leave the user in an unacceptable state with an non-functional app. Therefore a retry mechanism can only improve the user experience.
+
+## Impact on existing code
+
+The proposal doesn't change when the video or audio manager becomes streamable. The library should still listen for SDL Core information when the app becomes streamable. This proposal only synchronizes `StartService` requests and adds a retry mechanism to overcome failed attempts.
+
+All libraries require changes on the video and audio managers. Instead of actively sending `StartService` requests both managers have to wait for another source to trigger the starting routine. The implementation details are up to the project maintainers and the implementors. The iOS library already contains a general streaming manager called `SDLStreamingMediaManager` which could take the role of starting the video and audio manager. A similar manager could be added to the Java Suite taking over the responsibilities from the SDL manager.
+
+It is necessary to properly test that starting the handhshake process every two seconds. Numerous Ford IVIs were tested with one second of delay. Faster attempts caused SDL Core crashes on the IVI resulting in no app support whatsoever. It is possible that RPCs or a policy update can increase the stress as well and could cause crashes even with >1 second of delay. In rare cases, testing showed undefined behavior with one second of delay if the app is also sending voice commands or choice sets. More testing is needed to gain confidence of a faster retry mechanism. New head units should be tested against one second as well to ensure that the retry delay can be reduced in future. Other SDLC partners which use SDL security should also make sure the IVIs are supporting the retry mechanism.
+
+## Alternatives considered
+
+The SDL protocol implementation in the libraries can be synchronized with a lock, mutex or semaphore allowing only one secured `StartService` attempt. This method was not considered as the main solution because such locks generally block an entire thread and they require logic for cancellation. The main solution is considered to be more aligned with the SDL library implementation and the state machines.
+
+Alternatives to the infinite retry mechanism is to limit the retries to N times and increase the wait delay over time to avoid unnecessary stress to the infotainment system; for the cost of a good user experience and more delay.
```

Comment:
```suggestion
An alternative to the infinite retry mechanism is to limit the retries to N times and increase the wait delay over time to avoid unnecessary stress to the infotainment system. This alternative is at the cost of a good user experience and more delay.
```